### PR TITLE
Make AMD parser modules extensible again

### PR DESCRIPTION
**Pull request by chuckdumont**
**What kind of change does this PR introduce?** 

Structural (non-functional) changes to the AMD parser modules that restore extensibility which was lost in the v4 upgrade

**Did you add tests for your changes?**

No.  Not needed.  No functional changes.

**If relevant, link to documentation update:**

N/A

**Summary**

Prior to V4, code reuse in the AMD parser modules could be achieved through the use of the `call require:amd:array`, `call require:amd:item`, `call define:amd:array` and `call define:amd:item` plugin events.  These events were removed in V4, with no alternative to facilitate reuse of parser code by plugins.  This PR restores extensibility through the use of class inheritance (a cleaner approach for the use case).  The extensibility is needed by [dojo-webpack-plugin](https://github.com/OpenNTF/dojo-webpack-plugin) which needs to subclass and provide custom implementations of thie `processArray`  and `processItem` methods while leveraging the core parser functionality of the base classes.

**Does this PR introduce a breaking change?**

No

**Other information**

This PR is best reviewed by ignoring whitespace changes.

### CODE REVIEW

**Comment by chuckdumont** [comment]
Not sure what to do about the CLA not signed yet warning.  When I click on the link, it tell me that I've signed the CLA.

**Comment by chuckdumont** [comment]
Hi.  It's been 10 days since this PR was posted.  I someone going to review it or is there something process step I'm missing?

File: lib/dependencies/AMDDefineDependencyParserPlugin.js [hunk 1]
```diff
@@ -40,272 +40,276 @@ class AMDDefineDependencyParserPlugin {
...
-			range,
-			arrayRange,
-			functionRange,
-			objectRange,
-			namedModule
+	apply(parser) {
+		parser.hooks.call.for("define").tap(
+			"AMDDefineDependencyParserPlugin",
+			this.processCallDefine.bind(
+				Object.create(this, {
```
Comment (sokra) [inline lib/dependencies/AMDDefineDependencyParserPlugin.js]: This is weird. Just add the `parser` as additional argument to call member methods.

File: lib/dependencies/AMDRequireDependenciesBlockParserPlugin.js [hunk 2]
```diff
@@ -46,8 +46,116 @@ class AMDRequireDependenciesBlockParserPlugin {
...
+		);
+		if (!dep) return;
+		dep.loc = expr.loc;
+		dep.optional = !!this.parser.scope.inTry;
+		this.parser.state.current.addDependency(dep);
+		return true;
+	}
 
+	processCallRequire(expr) {
 		const processArrayForRequestString = param => {
```
Comment (sokra) [inline lib/dependencies/AMDRequireDependenciesBlockParserPlugin.js]: For consistency `processArrayForRequestString` and `processItemForRequestString` should also be member methods.

File: lib/dependencies/AMDDefineDependencyParserPlugin.js [hunk 3]
```diff
@@ -40,272 +40,276 @@ class AMDDefineDependencyParserPlugin {
...
+			dep.localModule = LocalModulesHelpers.addLocalModule(
+				parser.state,
+				namedModule
+			);
+		}
+		parser.state.current.addDependency(dep);
+		return true;
+	}
+
+	newDefineDependency(...args) {
```
Comment (sokra) [inline lib/dependencies/AMDDefineDependencyParserPlugin.js]: Write out all arguments explicitly.

File: lib/dependencies/AMDDefineDependencyParserPlugin.js [hunk 4]
```diff
@@ -40,272 +40,276 @@ class AMDDefineDependencyParserPlugin {
...
+		return new AMDDefineDependency(...args);
+	}
+	newRequireArrayDependency(...args) {
```
Comment (sokra) [inline lib/dependencies/AMDDefineDependencyParserPlugin.js]: Write out all arguments explicitly.

File: lib/dependencies/AMDDefineDependencyParserPlugin.js [hunk 5]
```diff
@@ -40,272 +40,276 @@ class AMDDefineDependencyParserPlugin {
...
+		return new AMDRequireArrayDependency(...args);
+	}
+	newRequireItemDependency(...args) {
```
Comment (sokra) [inline lib/dependencies/AMDDefineDependencyParserPlugin.js]: Write out all arguments explicitly.

File: lib/dependencies/AMDRequireDependenciesBlockParserPlugin.js [hunk 6]
```diff
@@ -70,172 +178,82 @@ class AMDRequireDependenciesBlockParserPlugin {
...
+				}
+			} finally {
+				parser.state.current = old;
+				if (dep) parser.state.current.addBlock(dep);
+			}
+			return true;
+		}
+	}
+
+	newRequireDependenciesBlock(...args) {
```
Comment (sokra) [inline lib/dependencies/AMDRequireDependenciesBlockParserPlugin.js]: Write out all arguments explicitly.

File: lib/dependencies/AMDRequireDependenciesBlockParserPlugin.js [hunk 7]
```diff
@@ -70,172 +178,82 @@ class AMDRequireDependenciesBlockParserPlugin {
...
+		return new AMDRequireDependenciesBlock(...args);
+	}
+	newRequireItemDependency(...args) {
```
Comment (sokra) [inline lib/dependencies/AMDRequireDependenciesBlockParserPlugin.js]: Write out all arguments explicitly.

File: lib/dependencies/AMDRequireDependenciesBlockParserPlugin.js [hunk 8]
```diff
@@ -70,172 +178,82 @@ class AMDRequireDependenciesBlockParserPlugin {
...
+		return new AMDRequireItemDependency(...args);
+	}
+	newRequireArrayDependency(...args) {
```
Comment (sokra) [inline lib/dependencies/AMDRequireDependenciesBlockParserPlugin.js]: Write out all arguments explicitly.

File: lib/dependencies/AMDRequireDependenciesBlockParserPlugin.js [hunk 9]
```diff
@@ -46,8 +46,116 @@ class AMDRequireDependenciesBlockParserPlugin {
 	}
 
 	apply(parser) {
-		const options = this.options;
+		parser.hooks.call.for("require").tap(
+			"AMDRequireDependenciesBlockParserPlugin",
+			this.processCallRequire.bind(
+				Object.create(this, {
```
Comment (sokra) [inline lib/dependencies/AMDRequireDependenciesBlockParserPlugin.js]: Same

**Comment by chuckdumont** [comment]
@sokra the requested changes have been made.  Thanks.

**Comment by sokra** [comment]
Thanks

**Pull request merged**